### PR TITLE
num_ivy_arrays_on_dev

### PR DESCRIPTION
**Pull request by davidlee1102**

### CODE REVIEW

**Comment by davidlee1102** [comment]
Hi Simone,
I have fixed this error, please take a look on that 🙇 
Thank you!
David Lee

**Comment by davidlee1102** [comment]
Hi Simone,
I did see that all the unsuccessful did not come from my code, could you recheck, I did see that was happened on others' file, not my code
Thank you!
David Lee

**Comment by simonetgordon** [comment]
Hi David,

Could you pull from master? Some fixes have been introduced for some of the tests, thanks!

**Comment by davidlee1102** [comment]
Hi Simone,
Please check my code again
Thank you!
David

**Comment by simonetgordon** [comment]
Hi David,

So currently your examples are failing the docstring tests. Have you tried running this function using the master version of ivy? When I do, I need to write the examples in a different manner to the way you have written them and obtain different results ( the # numbers is what was printed out)

```
import ivy #ivy master

ivy.set_backend('numpy')

x = ivy.array([-1,0,5.2])
y = ivy.num_ivy_arrays_on_dev(ivy.default_device())
print(y) # 1

x = ivy.array([-1,0,5.2])
y = ivy.array([-1,0,5.2,6])
z = ivy.num_ivy_arrays_on_dev(ivy.default_device())
print(z) # 2

x = ivy.Container(a=ivy.array([0., 1., 2.]), b=ivy.array([3., 4., 5.]))
z = ivy.num_ivy_arrays_on_dev(ivy.default_device())
print(z) # 3
```

Make sure to follow the [Contributer Guide](https://lets-unify.ai/ivy/deep_dive/11_docstring_examples.html#docstring-examples) closely, particularly for containers.

**Comment by davidlee1102** [comment]
Hi Simone, 
I am confused about the result came from the code running 
Here is the example I found and understood on ivy/master
`    Examples
    --------
    >>> x = ivy.array([-1,0,5.2])
    >>> y = ivy.dev(x)
    >>> z = ivy.num_ivy_arrays_on_dev(y)
    >>> print(z)
    2` 
And this is the code you were running and result:
`import ivy #ivy master

ivy.set_backend('numpy')

x = ivy.array([-1,0,5.2])
y = ivy.num_ivy_arrays_on_dev(ivy.default_device())
print(y) # 1

x = ivy.array([-1,0,5.2])
y = ivy.array([-1,0,5.2,6])
z = ivy.num_ivy_arrays_on_dev(ivy.default_device())
print(z) # 2

x = ivy.Container(a=ivy.array([0., 1., 2.]), b=ivy.array([3., 4., 5.]))
z = ivy.num_ivy_arrays_on_dev(ivy.default_device())
print(z) # 3`
About the result, I think somewhere was wrong, could you recheck on that? 🙇 
Thank you,
David Lee

**Comment by simonetgordon** [comment]
Hi David,

The original example was failing the tests too. My code was run on a python script importing the master branch of ivy so it's not wrong. Most likely, there have been changes to the function since the original example was written.

You can also run my code by making a new file in your local clone of ivy and running it on the command line/terminal.

**Comment by simonetgordon** [comment]
Hi David,

It would be best if you could run the code yourself and use this to write the docstring examples. Are you using PyCharm or VSCode to view and edit the ivy source code?

**Comment by davidlee1102** [comment]
HI Simone,
I will edit it follow on my code and example
About the coding ide, I am using Pycharm for coding Ivy
I will notify you after finishing the next commit
Thank you! 
David

**Comment by davidlee1102** [comment]
Hi Simone,
I have tried 
!pip install ivy-core

After that here is my code
`import ivy
import numpy


# Press the green button in the gutter to run the script.
if __name__ == '__main__':
    ivy.set_framework('numpy')
    x = ivy.array([-1, 0, 5.2])
    y = ivy.num_ivy_arrays_on_dev(ivy.default_device())
    print(y)`
However, the error was appeared; I just want to know how can you run the code?
Here is the error `AttributeError: module 'collections' has no attribute 'Hashable'`
Thank you,
David

**Comment by davidlee1102** [comment]
Hi Simone,
I have tried and the code  was running successful, however, I am confused about:
1. The name of issue was: num_arrays_on_dev | However the def name on device.py was num_ivy_arrays_on_dev, could you please check again
2. The read docs seem to be old, I have put ivy.set_framework('numpy') instead of ivy.set_backend('numpy'), It may give little bit confused
3. The result I have tried was the same with the examples, it was different with your result
I hope that you can check it and let me know about that
Thank you,
David

**Comment by simonetgordon** [comment]
Hi David,

1. the correct function is now `num_ivy_arrays_on_dev`. The version without 'ivy' was probably an old iteration of the function.
2. ` ivy.set_backend('numpy')` works in scripts which have imported ivy cloned recently from the master branch. If ` ivy.set_backend('numpy')` does not work for you, you are not importing the master branch ivy, but rather the stable ivy.

As I mentioned before, you can create a new .py script under the /ivy directory in PyCharm and `import ivy` in this script. As long as your local fork is up-to-date with the ivy master branch, you should be able to run ` ivy.set_backend('numpy')` and the code I have written above successfully.

Kind Regards, 

Simone

**Comment by davidlee1102** [comment]
Hi Simone,
I have fixed and It is currently running well, please check on my code 🙇 
About the result, I have found that your result was wrong in one function I have attached below
`
x = ivy.Container(a=ivy.array([0., 1., 2.]), b=ivy.array([3., 4., 5.]))
z = ivy.num_ivy_arrays_on_dev(ivy.default_device())
print(z) # 3
`
The result should be 2, not 3
I have also update for container and for both native_array and array too, hope that it can work well
Thank you for helping me to contribute to the ivy
Best regards,
David

**Comment by simonetgordon** [comment]
Hi David,

Glad you got it working! And indeed it should have been #2 instead of #3, I mistyped but did get 2 as the answer, my mistake.

Small thing to fix with the examples: the `set_backend` line is not needed, as it should be tested for all backends.

File: ivy/functional/ivy/device.py [hunk 1]
```diff
@@ -150,12 +150,148 @@ def num_ivy_arrays_on_dev(device: ivy.Device) -> int:
...
+    >>> print(y)
+    {
+        4
+    }
+
+    With a mix of :code:`ivy.Container` and :code:`ivy.NativeArray` input:
+
+    >>> ivy.set_framework('numpy')
+    >>> x = ivy.Container(x1= ivy.native_array([-1,0,5.2]))
+    >>> y = ivy.num_arrays_on_dev(ivy.default_device())
```
Comment (simonetgordon) [inline ivy/functional/ivy/device.py:259]: should be num_ivy_arrays_on_dev

File: ivy/functional/ivy/device.py [hunk 2]
```diff
@@ -150,12 +150,148 @@ def num_ivy_arrays_on_dev(device: ivy.Device) -> int:
...
+    >>> print(y)
+    {
+        0
+    }
+
+    >>> ivy.set_framework('numpy')
+    >>> x = ivy.Container(x1=ivy.native_array([-1, 0, 5.2, 6, 5.3]),\
+                        x2=ivy.native_array([-1, 0, 5.2, 4, 5]))
+    >>> y = ivy.num_arrays_on_dev(ivy.default_device())
```
Comment (simonetgordon) [inline ivy/functional/ivy/device.py:268]: should be num_ivy_arrays_on_dev

File: ivy/functional/ivy/device.py (hunk 1 above)
Comment (davidlee1102) [reply ivy/functional/ivy/device.py]: Thank you, I have fixed 🙇

File: ivy/functional/ivy/device.py (hunk 2 above)
Comment (davidlee1102) [reply ivy/functional/ivy/device.py]: Thank you, I have fixed 🙇

**Comment by davidlee1102** [comment]
Hi Simone,
I have fixed, so hope that it's finally for this function 😸 
Thank you for supporting me,
David

**Comment by simonetgordon** [comment]
Hi David,

There is the warning:
`UserWarning: Output is unequal to the docstrings output: num_ivy_arrays_on_dev`
appearing in the 'details' of the 'Run Docstring Tests'.

Please double check all the outputs to the functions is correct. Also, the format is important. The { } brackets in the last few examples may be causing this warning.

Kind Regards, 

Simone

**Comment by davidlee1102** [comment]
Hi Simeone,
I have checked all the function, all the examples were working well and the result of this was correct;
Could you recheck for me? I dont know where caused this warning 
Thank you,
David

**Comment by davidlee1102** [comment]
Hello @simonetgordon ,
I have check with the other guys, and I did see that he/she also had the same error on this PR
`UserWarning: Output is unequal to the docstrings output:`
So could you please recheck my code again
Thank you,
David

**Comment by simonetgordon** [comment]
Hi David,

Sorry, I don't follow what you mean. The user warning shouldn't be there. If there is, there is an issue with the docstring example not matching the output of the function when it's run. Could you please try removing the { } brackets in the code and see if that gets the test passing without a warning?

**Comment by davidlee1102** [comment]
Hi Simone, 
I mean that you can check other's PR
For example: https://github.com/unifyai/ivy/pull/1512, you can see that they also had the same error with me `(UserWarning: Output is unequal to the docstrings output:
)` on the appearing in the 'details' of the 'Run Docstring Tests'.
It was happened not just only me, I have tried, however, inside the ivy docs, they said that I need to put result in {} if I use the ivy.Container()
Can you check it for me again, thank you
David

**Comment by davidlee1102** [comment]
Hi Simone,
I have fixed, It was warning with {} as the same as you told me, thank you 😺 
Could you check that for me again
Thank you!
David

**Comment by simonetgordon** [comment]
Hi David,

That's great! Ready to merge 😊 Thanks again for the contribution.

Simone

**Pull request merged**